sdk/core/azure-core/src/cryptography/md5.cpp

```diff
@@ -16,156 +16,195 @@
-    hash.resize(md5Context->hashLength);
-    NTSTATUS status = BCryptFinishHash(
-        md5Context->hashHandle,
-        reinterpret_cast<PUCHAR>(&hash[0]),
-        static_cast<ULONG>(hash.size()),
-        0);
-    if (!BCRYPT_SUCCESS(status))
+    hash.resize(m_hashLength);
+    if (!NT_SUCCESS(
+            m_status = BCryptFinishHash(
+                m_hashHandle,
+                reinterpret_cast<PUCHAR>(&hash[0]),
+                static_cast<ULONG>(hash.size()),
+                0)))
     {
-      throw std::runtime_error("BCryptFinishHash failed");
+      throw std::runtime_error("BCryptFinishHash failed with code" + m_status);
     }
     return hash;
   }
 
-#elif defined(AZ_PLATFORM_POSIX)
-
-  Md5Hash::Md5Hash()
+public:
+  Md5BCrypt()
   {
-    MD5_CTX* md5Context = new MD5_CTX;
-    m_md5Context = md5Context;
-    MD5_Init(md5Context);
+    m_buffer.resize(GetMD5AlgorithmProvider().ContextSize);
+    m_hashLength = GetMD5AlgorithmProvider().HashLength;
+
+    if (!NT_SUCCESS(
+            m_status = BCryptCreateHash(
+                GetMD5AlgorithmProvider().Handle,
+                &m_hashHandle,
+                reinterpret_cast<PUCHAR>(&m_buffer[0]),
+                static_cast<ULONG>(m_buffer.size()),
+                nullptr,
+                0,
+                0)))
+    {
+      throw std::runtime_error("BCryptCreateHash failed with code" + m_status);
+    }
   }
 
-  Md5Hash::~Md5Hash()
+  ~Md5BCrypt()
   {
-    MD5_CTX* md5Context = static_cast<MD5_CTX*>(m_md5Context);
-    delete md5Context;
+    if (m_hashHandle)
+    {
+      BCryptDestroyHash(m_hashHandle);
+    }
   }
+};
 
-  void Md5Hash::OnAppend(const uint8_t* data, std::size_t length)
+} // namespace
+Azure::Core::Cryptography::Md5Hash::Md5Hash() : m_implementation(std::make_unique<Md5BCrypt>()) {}
+
+#elif defined(AZ_PLATFORM_POSIX)
+
+class Md5OpenSSL : public Azure::Core::Cryptography::Hash {
+private:
+  std::unique_ptr<MD5_CTX> m_context;
+
+  void OnAppend(const uint8_t* data, std::size_t length)
   {
-    MD5_CTX* md5Context = static_cast<MD5_CTX*>(m_md5Context);
-    MD5_Update(md5Context, data, length);
+    MD5_Update(m_context.get(), data, length);
   }
 
-  std::vector<uint8_t> Md5Hash::OnFinal(const uint8_t* data, std::size_t length)
+  std::vector<uint8_t> OnFinal(const uint8_t* data, std::size_t length)
   {
     OnAppend(data, length);
-    MD5_CTX* md5Context = static_cast<MD5_CTX*>(m_md5Context);
     unsigned char hash[MD5_DIGEST_LENGTH];
-    MD5_Final(hash, md5Context);
+    MD5_Final(hash, m_context.get());
     return std::vector<uint8_t>(std::begin(hash), std::end(hash));
   }
 
+public:
+  Md5OpenSSL()
+  {
+    m_context = std::make_unique<MD5_CTX>();
+    MD5_Init(m_context.get());
+  }
+};
+
+} // namespace
+Azure::Core::Cryptography::Md5Hash::Md5Hash() : m_implementation(std::make_unique<Md5OpenSSL>()) {}
 #endif
+
+namespace Azure { namespace Core { namespace Cryptography {
+  Md5Hash::~Md5Hash() {}
```

Comment:
What's this supposed to do?

Reply:
Defines the dttor as no-op.
The base class `Hash` defines a virtual dttor `virtual ~Hash() = default;` which obligated derived classes to define how the class is destroyed.
For the top level abstraction, we don't care about how it is destoyed because it only wraps a second layer abstraction which depends on the OS and where the dttor matters.